### PR TITLE
[Feature] [HunyuanImage3] Use FlashAttention for image denoising

### PR DESCRIPTION
**Pull request by nussejzz**
## Summary

- Replace 4D mask + `repeat_kv` attention with split FlashAttention for HunyuanImage3 denoising, eliminating GQA KV expansion (4x memory saving) and mask overhead
- Six-path dispatch covering step 1 / steps 2-50 × non-SP / SP / fallback:
  - **Steps 2-50, non-SP**: packed varlen FA (best) or two FA calls (fallback) — timestamp→text+ts, image→all
  - **Steps 2-50, SP**: maskless Ulysses SP Attention + local timestamp fix on rank 0
  - **Step 1, non-SP**: causal(text+ts) + full(image→text+ts+image), eoi zeroed out
  - **Step 1, SP**: text causal (local) + image SP (text as joint KV) + timestamp fix on rank 0
  - **Fallback**: original 4D mask path (fp32 or no flash_attn)
- Remove eoi token from KV cache — no token ever attends to eoi's KV (mask column all-False), and eoi's hidden state is discarded by `ragged_final_layer` via `masked_select(image_mask)`. Saves cache memory and eliminates dead computation in step 1
- Path selection is fully automatic based on dtype (fp16/bf16), flash_attn availability, and sp_size. Compatible with TP.

## Key Design Decisions

- **Maskless SP + local fix**: instead of passing 4D mask through Ulysses AllToAll (impossible without repeat_kv), run full attention in SP then fix timestamp output on rank 0 with a single Q_len=1 FA call. Zero extra memory, negligible overhead.
- **eoi is a ghost token**: verified in both original HunyuanImage-3.0 code and vllm-omni — `image_mask` marks only image token positions as True; eoi position is False. `ragged_final_layer` uses `masked_select(image_mask)` (step 1) or `x[:, 1:]` (steps 2+) to extract only image tokens. eoi computation in step 1 Path ④ replaced with `torch.zeros_like`.

## Test Plan

```
python3 benchmarks/bench_hunyuan_flash_attn.py \
  --model /data/HunyuanImage-3.0 \
  --stage-configs-path vllm_omni/model_executor/stage_configs/hunyuan_image3_moe_dit.yaml \
  --prompt "A brown and white dog is running on the grass" \
  --output-prefix output_bench \
  2>&1 | tee bench_3way.log
```

## Test Result

image

### Theoretical test value

image

### Actual test value

image

### Parameter Explanation

Stage | varlen | flash | sdpa
-- | -- | -- | --
prep/reshape — input tensor transform | reshape: 0.085ms | prep: 0.050ms | transpose×3 + .contiguous(): 0.053ms
repeat_kv — KV heads expansion (8→32) | None | None | 0.033ms
kernel — attention computation | 0.535ms | 0.555ms | 1.152ms
post — output transform | None | reshape: 0.014ms | transpose + reshape: 0.015ms
total | 0.622ms | 0.619ms | 1.253ms

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py [hunk 1]
```diff
@@ -943,18 +954,32 @@ def __call__(
...
 
-        query = query.transpose(1, 2).contiguous()
-        key = key.transpose(1, 2).contiguous()
-        value = value.transpose(1, 2).contiguous()
-
-        key = repeat_kv(key, repeat_num)
-        value = repeat_kv(value, repeat_num)
-
-        attention_mask = attention_mask.contiguous()
+        use_flash = not first_step and _HAS_FLASH_ATTN and query.dtype in (torch.float16, torch.bfloat16)
```
Comment (gcanlin) [inline vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py]: Why not use the unified `Attention`?

Comment (nussejzz) [reply vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py]: Good point. This PR is a first step toward full varlen FlashAttention support for image denoising. I want to validate that dropping the <eoi> token and using flash_attn_func(causal=False) produces numerically correct results, and measure the latency improvement.

**Review by Gaohan123 — COMMENTED** [review]
Please supplement performance test results

**Comment by nussejzz** [comment]
> Please supplement performance test results

OK~

**Comment by hsliuustc0106** [comment]
why it only helps improve a little?

**Comment by nussejzz** [comment]
> why it only helps improve a little?

From my point of view, it is because the 1.06x is end-to-end (including MoE FFN with 64 experts, VAE decode, etc.), not attention-only. Attention is a small fraction of per-step compute

**Comment by hsliuustc0106** [comment]
> > why it only helps improve a little?
> 
> From my point of view, it is because the 1.06x is end-to-end (including MoE FFN with 64 experts, VAE decode, etc.), not attention-only. Attention is a small fraction of per-step compute

so, can you profile it?

**Comment by nussejzz** [comment]
> > > why it only helps improve a little?
> > 
> > 
> > From my point of view, it is because the 1.06x is end-to-end (including MoE FFN with 64 experts, VAE decode, etc.), not attention-only. Attention is a small fraction of per-step compute
> 
> so, can you profile it?

Sure, I'll do further analysis and testing once the foreign exchange arrives tomorrow.😊❤️

**Review by lishunyang12 — COMMENTED** [review]
Implementation looks correct. Two things:

1. Use `_unwrap_flash_output` helper instead of the inline `isinstance(attn_output, tuple)` check — the codebase already has this utility.

2. Test results (latency + numerical accuracy vs SDPA path) should be provided before merge.

**Comment by nussejzz** [comment]
PTAL! 😊 @hsliuustc0106 @princepride  @Gaohan123 @gcanlin @lishunyang12

File: vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py [hunk 2]
```diff
@@ -943,19 +964,66 @@ def __call__(
...
+            query = query.transpose(1, 2).contiguous()
+            key = key.transpose(1, 2).contiguous()
+            value = value.transpose(1, 2).contiguous()
 
-        attention_mask = attention_mask.contiguous()
+            key = repeat_kv(key, repeat_num)
+            value = repeat_kv(value, repeat_num)
 
-        attn_output = F.scaled_dot_product_attention(query, key, value, attn_mask=attention_mask, dropout_p=0.0)
+            attention_mask = attention_mask.contiguous()
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py:1250]: can we reuse vllm_omni.diffusion.attention.layer.Attention?

File: vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py
Comment (princepride) [inline vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py:1]: LGTM

**Review by princepride — COMMENTED** [review]
LGTM, but I want know why you changed the yaml file?

File: vllm_omni/model_executor/stage_configs/hunyuan_image3_moe.yaml [hunk 3]
```diff
@@ -19,7 +19,7 @@ stage_args:
       model_arch: HunyuanImage3ForCausalMM
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/stage_configs/hunyuan_image3_moe.yaml:19]: if we change the name of yaml, would it affect the hy_image ar part?

File: vllm_omni/model_executor/stage_configs/hunyuan_image3_moe_dit.yaml [hunk 4]
```diff
@@ -7,7 +7,7 @@ stage_args:
   - stage_id: 0
     stage_type: diffusion
     runtime:
-      devices: "0,1,2,3"
+      devices: "0,1,2,3,4,5,6,7" # Visible devices for this stage (CUDA_VISIBLE_DEVICES/torch.cuda.set_device)
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/stage_configs/hunyuan_image3_moe_dit.yaml]: I do not think we should change this yaml? you can add comment here: # for A100, please use at least xx devices

File: vllm_omni/model_executor/stage_configs/hunyuan_image3_moe.yaml [hunk 5]
```diff
@@ -19,7 +19,7 @@ stage_args:
...
       worker_cls: vllm_omni.worker.gpu_ar_worker.GPUARWorker
       scheduler_cls: vllm_omni.core.sched.omni_ar_scheduler.OmniARScheduler
-      gpu_memory_utilization: 0.3
+      gpu_memory_utilization: 0.5
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/stage_configs/hunyuan_image3_moe.yaml]: do not change the default yaml

File: vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py
Comment (nussejzz) [reply vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py:1]: The yaml changes are for consistency between the two stage configs:

Rename hunyuan_image_3_moe.yaml → hunyuan_image3_moe.yaml: align naming with hunyuan_image3_moe_dit.yaml (remove extra underscore before 3)
gpu_memory_utilization: 0.3 → 0.5: the original 0.3 causes KV cache OOM on both L40S-48G and A100-80G (model weights alone take 22.55 GiB/card with TP8, leaving no room for KV cache at 0.3)
DiT config: TP 4→8, devices 4→8: align with the AR config for 8-GPU deployment

File: vllm_omni/model_executor/stage_configs/hunyuan_image3_moe_dit.yaml (hunk 4 above)
Comment (nussejzz) [reply vllm_omni/model_executor/stage_configs/hunyuan_image3_moe_dit.yaml]: ok

File: vllm_omni/model_executor/stage_configs/hunyuan_image3_moe.yaml (hunk 5 above)
Comment (nussejzz) [reply vllm_omni/model_executor/stage_configs/hunyuan_image3_moe.yaml]: ok

File: vllm_omni/model_executor/stage_configs/hunyuan_image3_moe.yaml (hunk 3 above)
Comment (nussejzz) [reply vllm_omni/model_executor/stage_configs/hunyuan_image3_moe.yaml:19]: No, the yaml filename is not hardcoded anywhere in the codebase. It's always passed by the user via --stage-configs-path argument. The rename is purely for naming consistency with hunyuan_image3_moe_dit.yaml.

File: vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py
Comment (nussejzz) [reply vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py:1]: Haha, I deleted them all. I noticed that the newly merged YAML file used the same naming convention as before.😂

**Review by princepride — APPROVED** [review]
Good job!

File: vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py [hunk 6]
```diff
@@ -943,19 +945,37 @@ def __call__(
...
+            # Reuse the diffusion Attention layer for steps 2-50.
+            # Backend selection (FlashAttn/SDPA/etc.) is handled automatically
+            # via DIFFUSION_ATTENTION_BACKEND env var. USP is supported natively.
+            # Exclude last cached token from KV — image tokens cannot attend to it.
+            attn_output = self.attn_layer(
+                query,
+                key[:, :-1],
+                value[:, :-1],
+                attn_metadata=None,
+            )
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py:1145]: Actually, for steps 2–50, the attention is still not full. The query consists of 1 timestamp token and the image tokens, so the attention mask should be structured as follows:

Timestamp token row: [true, true, ..., false, false]
Image token rows: [true, true, ..., true, true]

Could you please double-check this?

**Review by princepride — CHANGES_REQUESTED** [review]
We need consider timestamp token's attention😂

**Comment by nussejzz** [comment]
> We need consider timestamp token's attention😂

No problem

File: vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py (hunk 6 above)
Comment (nussejzz) [reply vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py:1145]: Yes, I misunderstood the attention interval of the timestamp, but you can directly perform a full attn because its length is one.
I will fix it ASAP😊

File: vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py [hunk 7]
```diff
@@ -1042,37 +1051,216 @@ def __call__(
...
-        if self.sp_size > 1:
-            joint_text_key = repeat_kv(joint_text_key, repeat_num)
-            joint_text_value = repeat_kv(joint_text_value, repeat_num)
-
-        attention_mask = attention_mask.contiguous()
+        _flash_ok = query.dtype in (torch.float16, torch.bfloat16)
+        _has_fa = _flash_attn_func is not None
+        _has_varlen = _flash_attn_varlen_func is not None
+
+        if not first_step and _flash_ok and self.sp_size <= 1 and _has_varlen:
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py:1068]: need to rm https://github.com/vllm-project/vllm-omni/blob/1ca942999ab929af306297d8853d317b9c975896/vllm_omni/diffusion/models/hunyuan_image_3/pipeline_hunyuan_image_3.py#L92 and test

Comment (nussejzz) [reply vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py:1068]: ok

**Comment by lishunyang12** [comment]
Can you profile it using torch profiler ? Userful links: https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/profiling/

**Review by hsliuustc0106 — COMMENTED** [review]
## 🔍 PR #1975 Review: FlashAttention for HunyuanImage3 Denoising

### ✅ **Passed Items**
- Correctness, performance, implementation quality all good
- 6-path dispatch logic is correct
- EOI handling is reasonable

### ⚠️ **Items to Add**

**1. Functional Correctness Tests** (Required)
- Verify numerical consistency between FlashAttention and SDPA paths
- Cover all 6 dispatch paths

**2. E2E Tests** (Strongly Recommended)
- Verify actual image generation quality
- Compare outputs with fixed seed

### 📝 **Recommendations**

Before merging, please add:

```bash
# 1. Functional correctness tests
tests/diffusion/models/hunyuan_image_3/test_flash_attention.py
```
- Verify numerical consistency across all 6 dispatch paths with SDPA baseline

```bash
# 2. E2E tests
tests/e2e/test_hunyuan_image3_flash_attn.py
```
- Verify actual image generation quality
- Compare Flash and SDPA outputs with fixed seed

**Code quality is excellent, implementation logic is correct. Ready to merge after adding tests.** 👍

**Comment by nussejzz** [comment]
## Functional Correctness Test

Clipboard_Screenshot_1776085450
